src/serde/room_state.rs

```diff
@@ -0,0 +1,108 @@
+//! A module to deserialize a RoomState struct from incorrectly specified v1
+//! send_join endpoint.
+//!
+//! For more information, see this [GitHub issue](https://github.com/matrix-org/matrix-doc/issues/2541).
+
+use std::fmt;
+
+use serde::{
+    de::{self, Deserializer, Error, SeqAccess, Visitor},
+    ser::{SerializeSeq, Serializer},
+};
+
+use crate::v1::join::create_join_event::RoomState;
+
+pub fn serialize<S>(room_state: &RoomState, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let mut seq = serializer.serialize_seq(Some(2))?;
+    seq.serialize_element(&200)?;
+    seq.serialize_element(room_state)?;
+    seq.end()
+}
+
+pub fn deserialize<'de, D>(deserializer: D) -> Result<RoomState, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_seq(RoomStateVisitor)
+}
+
+struct RoomStateVisitor;
+
+impl<'de> Visitor<'de> for RoomStateVisitor {
+    type Value = RoomState;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("Room State response")
+    }
+
+    fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
```

Comment:
Maybe rename `V` to `A` like in the trait? Or are we already using `V` as the name for `SeqAccess` / `MapAccess` types elsewhere?

```suggestion
    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
```

Reply:
I think that `V` either popped up in rust-analyzer or in the documentation I was reading, so I just went for it. I'll change it.